src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -997,8 +997,15 @@ internal object GetParsedValues(HeaderDescriptor descriptor)
             return info.ParsedValue;
         }
 
+        internal virtual bool IsAllowedHeaderName(string headerName) => true;
+
         private void PrepareHeaderInfoForAdd(HeaderDescriptor descriptor, out HeaderStoreItemInfo info, out bool addToStore)
         {
+            if (!IsAllowedHeaderName(descriptor.Name))
+            {
+                throw new InvalidOperationException(SR.net_http_headers_not_allowed_header_name);
```

Comment:
We should modify `SR.net_http_headers_not_allowed_header_name` so that the headerName shows up as part of the exception text.

Instead of:
"Misused header name. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects."

it would be better to see, i.e.
"Misused header name, 'MY-HEADER-NAME'. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects."

Doing this will greatly help with diagnostics.

You'll need to modify `SR.net_http_headers_not_allowed_header_name` to add a `{0}` in the resource string and fix up other places this string resource is used.